### PR TITLE
decode: add Cisco HDLC decoder - v4

### PR DESCRIPTION
**Pull request by jasonish**
Previous PR:
https://github.com/inliniac/suricata/pull/2166

Redmine issue: https://redmine.openinfosecfoundation.org/issues/1807

Changes in this PR:
- Rebase required.

Cisco HDLC decoder. Support added for PCAP, ERF and DAG.

Prscript:
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/297
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/302

### CODE REVIEW

File: src/source-erf-file.c [hunk 1]
```diff
@@ -287,7 +305,19 @@ DecodeErfFile(ThreadVars *tv, Packet *p, void *data, PacketQueue *pq, PacketQueu
...
-    DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+    switch (p->datalink) {
+        case LINKTYPE_ETHERNET:
+            DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+            break;
+        case LINKTYPE_CHDLC:
+            DecodeCHDLC(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
+            break;
+        default:
+            SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED,
```
Comment (inliniac) [inline src/source-erf-file.c]: this should probably either be a fatal error or a debug message as this will print a line per packet

**Comment by inliniac** [comment]
Needs rebase.

File: src/source-erf-file.c (hunk 1 above)
Comment (jasonish) [inline src/source-erf-file.c]: I think this is consistent with af-packet and pcap though.

**Comment by inliniac** [comment]
On 30-09-16 21:14, Jason Ish wrote:
> *@jasonish* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/source-erf-file.c <https://github.com/inliniac/suricata/pull/2206>:
> 
>> @@ -287,7 +305,19 @@ DecodeErfFile(ThreadVars *tv, Packet *p, void *data, PacketQueue *pq, PacketQueu
>      /* Update counters. */
>      DecodeUpdatePacketCounters(tv, dtv, p);
>  
> -    DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> +    switch (p->datalink) {
> +        case LINKTYPE_ETHERNET:
> +            DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> +            break;
> +        case LINKTYPE_CHDLC:
> +            DecodeCHDLC(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> +            break;
> +        default:
> +            SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED,
> 
> I think this is consistent with af-packet and pcap though.

Care to fix those too? :)

**Comment by jasonish** [comment]
On Fri, Sep 30, 2016 at 1:17 PM, Victor Julien <notifications@github.com>
wrote:

> On 30-09-16 21:14, Jason Ish wrote:
> > *@jasonish* commented on this pull request.
> >
> > ------------------------------------------------------------------------
> >
> > In src/source-erf-file.c <https://github.com/inliniac/suricata/pull/2206
> >:
> >
> >> @@ -287,7 +305,19 @@ DecodeErfFile(ThreadVars *tv, Packet *p, void
> *data, PacketQueue *pq, PacketQueu
> > /* Update counters. */
> > DecodeUpdatePacketCounters(tv, dtv, p);
> >
> > - DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> > + switch (p->datalink) {
> > + case LINKTYPE_ETHERNET:
> > + DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> > + break;
> > + case LINKTYPE_CHDLC:
> > + DecodeCHDLC(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> > + break;
> > + default:
> > + SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED,
> >
> > I think this is consistent with af-packet and pcap though.
>
> Care to fix those too? :)


I really don't want it to be debug, as I really want to know - so I'm
thinking FatalError()?

**Comment by inliniac** [comment]
On 30-09-16 21:19, Jason Ish wrote:
> On Fri, Sep 30, 2016 at 1:17 PM, Victor Julien <notifications@github.com>
> wrote:
> 
>> On 30-09-16 21:14, Jason Ish wrote:
>> > *@jasonish* commented on this pull request.
>> >
>> > ------------------------------------------------------------------------
>> >
>> > In src/source-erf-file.c <https://github.com/inliniac/suricata/pull/2206
>> >:
>> >
>> >> @@ -287,7 +305,19 @@ DecodeErfFile(ThreadVars *tv, Packet *p, void
>> *data, PacketQueue *pq, PacketQueu
>> > /* Update counters. */
>> > DecodeUpdatePacketCounters(tv, dtv, p);
>> >
>> > - DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
>> > + switch (p->datalink) {
>> > + case LINKTYPE_ETHERNET:
>> > + DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
>> > + break;
>> > + case LINKTYPE_CHDLC:
>> > + DecodeCHDLC(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
>> > + break;
>> > + default:
>> > + SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED,
>> >
>> > I think this is consistent with af-packet and pcap though.
>>
>> Care to fix those too? :)
> 
> 
> I really don't want it to be debug, as I really want to know - so I'm
> thinking FatalError()?

Yeah. Although we really should check this at setup as well. But can't
hurt to have both.

**Comment by jasonish** [comment]
On Fri, Sep 30, 2016 at 1:21 PM, Victor Julien <notifications@github.com>
wrote:

> On 30-09-16 21:19, Jason Ish wrote:
> > On Fri, Sep 30, 2016 at 1:17 PM, Victor Julien <notifications@github.com
> >
> > wrote:
> >
> >> On 30-09-16 21:14, Jason Ish wrote:
> >> > *@jasonish* commented on this pull request.
> >> >
> >> > ------------------------------------------------------------
> ------------
> >> >
> >> > In src/source-erf-file.c <https://github.com/inliniac/
> suricata/pull/2206
> >> >:
> >> >
> >> >> @@ -287,7 +305,19 @@ DecodeErfFile(ThreadVars *tv, Packet *p, void
> >> *data, PacketQueue *pq, PacketQueu
> >> > /* Update counters. */
> >> > DecodeUpdatePacketCounters(tv, dtv, p);
> >> >
> >> > - DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> >> > + switch (p->datalink) {
> >> > + case LINKTYPE_ETHERNET:
> >> > + DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> >> > + break;
> >> > + case LINKTYPE_CHDLC:
> >> > + DecodeCHDLC(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
> >> > + break;
> >> > + default:
> >> > + SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED,
> >> >
> >> > I think this is consistent with af-packet and pcap though.
> >>
> >> Care to fix those too? :)
> >
> >
> > I really don't want it to be debug, as I really want to know - so I'm
> > thinking FatalError()?
>
> Yeah. Although we really should check this at setup as well. But can't
> hurt to have both.


Easy with pcap. Not easy with DAG, and doesn't look easy with af-packet.
So I know that pcap, and dag will get just a single link type, so doing a
fatal error is probably fine.  But af-packet, any chance it could a variety
of link types?

**Comment by inliniac** [comment]
On 30-09-16 21:26, Jason Ish wrote:
> On Fri, Sep 30, 2016 at 1:21 PM, Victor Julien <notifications@github.com>
> wrote:
> 
>> On 30-09-16 21:19, Jason Ish wrote:
>> > On Fri, Sep 30, 2016 at 1:17 PM, Victor Julien <notifications@github.com
>> >
>> > wrote:
>> >
>> >> On 30-09-16 21:14, Jason Ish wrote:
>> >> > *@jasonish* commented on this pull request.
>> >> >
>> >> > ------------------------------------------------------------
>> ------------
>> >> >
>> >> > In src/source-erf-file.c <https://github.com/inliniac/
>> suricata/pull/2206
>> >> >:
>> >> >
>> >> >> @@ -287,7 +305,19 @@ DecodeErfFile(ThreadVars *tv, Packet *p, void
>> >> *data, PacketQueue *pq, PacketQueu
>> >> > /* Update counters. */
>> >> > DecodeUpdatePacketCounters(tv, dtv, p);
>> >> >
>> >> > - DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
>> >> > + switch (p->datalink) {
>> >> > + case LINKTYPE_ETHERNET:
>> >> > + DecodeEthernet(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
>> >> > + break;
>> >> > + case LINKTYPE_CHDLC:
>> >> > + DecodeCHDLC(tv, dtv, p, GET_PKT_DATA(p), GET_PKT_LEN(p), pq);
>> >> > + break;
>> >> > + default:
>> >> > + SCLogError(SC_ERR_DATALINK_UNIMPLEMENTED,
>> >> >
>> >> > I think this is consistent with af-packet and pcap though.
>> >>
>> >> Care to fix those too? :)
>> >
>> >
>> > I really don't want it to be debug, as I really want to know - so I'm
>> > thinking FatalError()?
>>
>> Yeah. Although we really should check this at setup as well. But can't
>> hurt to have both.
> 
> 
> Easy with pcap. Not easy with DAG, and doesn't look easy with af-packet.
> So I know that pcap, and dag will get just a single link type, so doing a
> fatal error is probably fine. But af-packet, any chance it could a variety
> of link types?

I think it would have come up already. Error printing at line rate
should bring users to mailinglist/redmine, and AFAIK it hasn't been
reported. So lets go with fatal error.

**Pull request closed without merging**